### PR TITLE
Feature/estimator

### PR DESCRIPTION
**Pull request by antschum**
**Description of changes**
create_estimator from sklearn; improved typing

### CODE REVIEW

File: augurpy/estimators.py [hunk 1]
```diff
@@ -1,25 +1,57 @@
-from typing import Dict, List, Optional, Union
+from typing import Dict, Literal, Union
 
 from sklearn.ensemble import RandomForestClassifier, RandomForestRegressor
 from sklearn.linear_model import LogisticRegression
+from typing_extensions import TypedDict
+
+
+class Params(TypedDict, total=False):
```
Comment (Zethson) [inline augurpy/estimators.py:11]: Could this be a Dataclass?

File: augurpy/estimators.py [hunk 2]
```diff
@@ -1,25 +1,57 @@
...
+    """Type signature for random forest and logistic regression parameters."""
+
+    n_estimators: int
+    max_depth: int
+    max_features: Union[Literal["auto"], Literal["log2"], Literal["sqrt"], int, float]
+    penalty: Union[Literal["l1"], Literal["l2"], Literal["elasticnet"], Literal["none"]]
```
Comment (Zethson) [inline augurpy/estimators.py:17]:
```suggestion
    penalty: Union[Literal["l1"], Literal["l2"], Literal["elasticnet"], None
```

Think this should also work? We can handle None.

File: augurpy/estimators.py [hunk 3]
```diff
@@ -1,25 +1,57 @@
...
         classifier: classifier to use in calculating the area under the curve either random forest or logistic regression
-        rf_params: list of parameters for random forest
-        lr_params: list of parameters for logistic regression
+        params: list of parameters for estimator
 
     Returns:
         Estimator object that classifies.
     """
-    pass
+    if classifier == "rf" and classification:
```
Comment (Zethson) [inline augurpy/estimators.py]: To make this scalable I like switch statements https://github.com/mikeckennedy/python-switch or Python 3.10 match statements (think we cannot go to Python 3.10 yet)

We might add e.g. Gradient Boosted Trees as a third option later.

File: augurpy/estimators.py [hunk 4]
```diff
@@ -1,25 +1,57 @@
...
+        estimator = RandomForestClassifier(
+            n_estimators=params.get("n_estimators", 100),
+            max_depth=params.get("max_depth", None),
+            max_features=params.get("max_features", "auto"),
+        )
+        print(estimator, " Object created.")
```
Comment (Zethson) [inline augurpy/estimators.py]: Remove, please.

File: augurpy/estimators.py [hunk 5]
```diff
@@ -1,25 +1,57 @@
...
+        estimator = RandomForestRegressor(
+            n_estimators=params.get("n_estimators", 100),
+            max_depth=params.get("max_depth", None),
+            max_features=params.get("max_features", "auto"),
+        )
+        print(estimator, " Object created.")
+
+    elif classifier == "lf" and not classification:
+        estimator = LogisticRegression()
+        print(estimator + " Object created.")
```
Comment (Zethson) [inline augurpy/estimators.py]: Remove, please.

File: augurpy/estimators.py [hunk 6]
```diff
@@ -1,25 +1,57 @@
...
+
+    else:
+        raise Exception("Missing valid input.")
```
Comment (Zethson) [inline augurpy/estimators.py]: Always try to offer a resolution. "Missing valid input. Valid input is xyz".

**Review by Zethson — CHANGES_REQUESTED** [review]
A couple of requests.
Don't forget that you now have to merge development into this branch.

File: augurpy/estimators.py (hunk 1 above)
Comment (antschum) [reply augurpy/estimators.py:11]: Yes! The decorator you mean, right? Done.

File: augurpy/estimators.py (hunk 3 above)
Comment (antschum) [reply augurpy/estimators.py]: Yess I like those too! Done.

File: augurpy/estimators.py (hunk 2 above)
Comment (antschum) [reply augurpy/estimators.py:17]: Done.

Comment (antschum) [reply augurpy/estimators.py:17]: Ah but this goes directly into the [logistic regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html?highlight=logistic%20regression#sklearn.linear_model.LogisticRegression) as a parameter. It needs 'none' not None, so I would rather keep the string, unless Im missing something.

File: augurpy/estimators.py (hunk 4 above)
Comment (antschum) [reply augurpy/estimators.py]: done.

File: augurpy/estimators.py (hunk 5 above)
Comment (antschum) [reply augurpy/estimators.py]: done

File: augurpy/estimators.py (hunk 6 above)
Comment (antschum) [reply augurpy/estimators.py]: done

File: augurpy/estimators.py (hunk 2 above)
Comment (Zethson) [reply augurpy/estimators.py:17]: Huh, didn't remember this. But you are right. Let's do it like you had it!

File: augurpy/estimators.py (hunk 1 above)
Comment (Zethson) [reply augurpy/estimators.py:11]: Yes, the decorator. The params object looks like it's purely a data container but does not need any functionality.

File: augurpy/estimators.py [hunk 7]
```diff
@@ -1,25 +1,63 @@
...
+class Params(TypedDict, total=False):
+    """Type signature for random forest and logistic regression parameters."""
+
+    n_estimators: int
+    max_depth: int
+    max_features: Union[Literal["auto"], Literal["log2"], Literal["sqrt"], int, float]
+    penalty: Union[Literal["l1"], Literal["l2"], Literal["elasticnet"], Literal["none"]]
+
+
+def _raise_exception():
```
Comment (Zethson) [inline augurpy/estimators.py]: I don't see the need for this method because it's a single line anyways.

File: augurpy/estimators.py [hunk 8]
```diff
@@ -1,25 +1,63 @@
...
         classifier: classifier to use in calculating the area under the curve either random forest or logistic regression
-        rf_params: list of parameters for random forest
-        lr_params: list of parameters for logistic regression
+        params: list of parameters for estimator
 
     Returns:
         Estimator object that classifies.
     """
-    pass
+    definition = classifier + " categorical: " + str(classification)
```
Comment (Zethson) [inline augurpy/estimators.py]: definition is kind of an ambiguous name, no? If I read this without context I do not know what this is supposed to mean. Can we find a better name?

Comment (Zethson) [reply augurpy/estimators.py]: Also, can we make this an f string?

File: augurpy/estimators.py [hunk 9]
```diff
@@ -1,25 +1,63 @@
...
+    with switch(definition) as c:
+        c.case(
+            "rf categorical: True",
+            lambda: RandomForestClassifier(
+                n_estimators=params.get("n_estimators", 100),
```
Comment (Zethson) [inline augurpy/estimators.py:61]: I'd suggest to document the default parameters maybe in the docstring of this function?

**Review by Zethson — CHANGES_REQUESTED** [review]
A couple more minor requests :) 
But almost there!

File: augurpy/estimators.py (hunk 7 above)
Comment (antschum) [reply augurpy/estimators.py]: Its difficult to raise an exception directly in the lambda statement ([here](https://stackoverflow.com/questions/8294618/define-a-lambda-expression-that-raises-an-exception)), thats why I have the extra function. Do you like one of the other options more? I could also think of a different default for the switch statement I guess.

Comment (Zethson) [reply augurpy/estimators.py]: Ahh, didn't notice this.

Clear favorite:
```
type(lambda: 0)(type((lambda: 0).__code__)(
    1,0,1,1,67,b'|\0\202\1\0',(),(),('x',),'','',1,b''),{}
)(Exception())
```

(Please keep it like you had it) :)

File: augurpy/estimators.py (hunk 8 above)
Comment (antschum) [reply augurpy/estimators.py]: Hurray name finding. How about estimator_string? define_classifier? define_estimator? match_estimator? switch_estimator?
I think I like match estimator most, but switch_estimator might be less confusing🤷‍♀️

Comment (Zethson) [reply augurpy/estimators.py]: I think we should implement it differently. Simply include more options for the `classifier` parameter: Can be one of `random_forest_regressor`, `random_forest_classifier`, [...].
Further, the classification parameter is currently not documented :)
Could you also expand a little bit on the 1 sentence description of the docstring of this method?

Something more along the lines of "Creates a model object of the provided type and populates it with desired parameters".

File: augurpy/estimators.py (hunk 9 above)
Comment (antschum) [reply augurpy/estimators.py:61]: Do you mean the docstring of the Params class?

Comment (Zethson) [reply augurpy/estimators.py:61]: No, of the `create_estimator` method I thought.

Comment (antschum) [reply augurpy/estimators.py:61]: I also noticed that the Params does not do what I had hoped: I would like for it to only let those defined keys be generated and to throw an error or something similar when something undefined is passed.

Comment (antschum) [reply augurpy/estimators.py:61]: I think the params thing is another topic - ill ask again tomorrow :)

Comment (Zethson) [reply augurpy/estimators.py:61]: Yeah, we can discuss tomorrow. :+1:

Comment (antschum) [reply augurpy/estimators.py:61]: > No, of the `create_estimator` method I thought.

Is there a specific style I can orient this on?

Comment (Zethson) [reply augurpy/estimators.py:61]: > > No, of the `create_estimator` method I thought.
> 
> Is there a specific style I can orient this on?

Nah. But it's nothing major. Just have an attempt :)

File: augurpy/estimators.py [hunk 10]
```diff
@@ -70,7 +72,9 @@ def create_estimator(
             ),
         )
         c.case("logistic_regression_classifier", lambda: LogisticRegression(penalty=params.get("penalty", "l2")))
-        c.default(_raise_exception)
+        c.default(_raise_exception)(
+            "Missing valid input. Choose rf or lr for categorical labels and rf for continuous labels."
+        )
```
Comment (antschum) [inline augurpy/estimators.py:77]: Das hier funktioniert so leider nicht, ich bekomme da immer einen `TypeError: 'NoneType' object is not callable`. Das folgende scheint allerdings zu funktionieren: 
`c.default(lambda: _raise_exception("Missing valid input"))`
Oder lieber die vorherige version?

Comment (Zethson) [reply augurpy/estimators.py:77]: `c.default(lambda: _raise_exception("Missing valid input"))`

This looks fine to me.

**Pull request merged**